apstools/devices/motor_factory.py

```diff
@@ -0,0 +1,136 @@
+"""
+Motor Bundle Factory
+====================
+
+Factory for creating bundles with any number of positioners.
+
+.. autosummary::
+
+    ~axis_component
+    ~mb_class_factory
+    ~mb_creator
+
+.. rubric:: Examples
+
+.. code-block:: python
+    :linenos:
+
+    xy_stage = mb_creator(prefix="IOC:", motors={"x": "m21", "y": "m22"})
+
+.. seealso:: :doc:`How to use 'mb_creator()' </examples/ho_motor_factory>`
+"""
+
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Type
+from typing import Union
+
+from ophyd import Component
+from ophyd import EpicsMotor
+from ophyd import Kind
+from ophyd import MotorBundle
+from ophyd import SoftPositioner
+
+from apstools.utils import dynamic_import
+
+DEFAULT_DEVICE_BASE_CLASSES: List[Union[str, Type]] = [MotorBundle]
+DEFAULT_DEVICE_CLASS_NAME: str = "MB_Device"
+DEFAULT_MOTOR_LABELS: List[str] = ["motors"]
+DEFAULT_KIND: Kind = Kind.hinted | Kind.normal
+
+MOTORS_TYPE = Union[List[str], Dict[str, Optional[Union[str, Dict]]]]
+
+
+def axis_component(parms: dict[str, Any]) -> Component:
+    """Return a Component with 'parms for this axis."""
+    parms = parms or {}
+    if isinstance(parms, str):
+        parms = dict(prefix=parms)
+    attrs = dict(kind=DEFAULT_KIND, labels=DEFAULT_MOTOR_LABELS)
+    has_prefix: bool = parms.get("prefix") is not None
+    axis_class_name = parms.pop("class", None)
+    factory = parms.pop("factory", None)
+    if axis_class_name is not None and factory is not None:
+        raise ValueError(f"Unsupported configuration: {axis_class_name=!r} and {factory=!r}")
+    if factory is not None:
+        creator = factory.pop("function")
+        if isinstance(creator, str):
+            creator = dynamic_import(creator)
+        base = creator(**factory)
+        # Add 'prefix' if missing?
+    elif axis_class_name is None:
+        base = EpicsMotor if has_prefix else SoftPositioner
+        if base == EpicsMotor and "prefix" not in parms:
+            parms["prefix"] = ""
+    else:
+        # Allow for any positioner type.
+        base = dynamic_import(axis_class_name)
+    attrs.update(**parms)
+
+    args = [base]
+    prefix = attrs.pop("prefix", None)
+    if prefix is not None:
+        args.append(prefix)
+    return Component(*args, **attrs)
+
+
+def mb_class_factory(
+    motors: MOTORS_TYPE = None,
```

Comment:
I don't think `None` is valid as a `MOTORS_TYPE`. Maybe `MOTORS_TYPE | None`?

Reply:
done